Email notifications

simong opened the pull request:
This will send out daily email notifications by using the SendGrid API.
Currently it covers:
 * New comments on your assets
 * New replies on your comments
 * New chat messages on your whiteboards

It also sends an email when somebody gets added to a whiteboard

simong commented:
I forgot to mention, these are the required DB migrations:

```
alter table users add canvas_email character varying(255);
alter table courses add assetlibrary_url character varying(255);
alter table courses add whiteboards_url character varying(255);
alter table canvas add name character varying(255);
alter table canvas add logo character varying(255);
```

nicolaasmatthijs commented on config/default.json:
```diff
@@ -32,7 +33,12 @@
   },
   "canvasPoller": {
-    "enabled": true,
+    "enabled": false,
```
I think the tests already disable the poller, so this can probably remain enabled by default?

nicolaasmatthijs commented on email-templates/_shared/style.css:
```diff
@@ -0,0 +1,349 @@
+    border: 0;
+}
+
+.muted {
```
It doesn't look like this is being used

nicolaasmatthijs commented on email-templates/_shared/style.css:
```diff
@@ -0,0 +1,349 @@
+    margin: 40px auto 0;
+}
+
+#logo td {
```
These logo and greeting styles should be in the `header` section

nicolaasmatthijs commented on email-templates/_shared/style.css:
```diff
@@ -0,0 +1,349 @@
+    width: 300px;
+}
+
+.preview-tile-container.collabdoc {
```
This is unused

nicolaasmatthijs commented on email-templates/_shared/style.css:
```diff
@@ -0,0 +1,349 @@
+    background-position: top left;
+}
+
+.preview-tile-container-wide {
```
This is also unused

nicolaasmatthijs commented on email-templates/_shared/style.css:
```diff
@@ -0,0 +1,349 @@
+}
+
+.preview-asset-metadata {
+    color: #aeaaaa;
```
Should be capital case

nicolaasmatthijs commented on email-templates/_shared/style.css:
```diff
@@ -0,0 +1,349 @@
+    width: 100%;
+}
+
+.activity-row .activity-thumbnail-container {
```
This doesn't seem to be used

nicolaasmatthijs commented on email-templates/_shared/style.css:
```diff
@@ -0,0 +1,349 @@
+    margin-left: 40px;
+}
+
+.activity-row .activity-comment-container.comment-level-2 {
```
`comment-level-2` doesn't appear to be used

nicolaasmatthijs commented on email-templates/_shared/style.css:
```diff
@@ -0,0 +1,349 @@
+
+/* Markdown formatting */
+
+.activity-row .activity-comment-container .activity-comment .activity-comment-message h1 {
```
Comments and chat messages don't have any Markdown formatting. The only formatting that should be here is converting URLs to links.

nicolaasmatthijs commented on email-templates/_shared/style.css:
```diff
@@ -0,0 +1,349 @@
+    border: 1px solid #00628E;
+    color: #FFF;
+    text-decoration: none;
+    padding: 15px 20px;
```
Alphabetical

nicolaasmatthijs commented on email-templates/_shared/style.css:
```diff
@@ -0,0 +1,349 @@
+/**
+ * Invitations
+ */
+
```
This doesn't appear to be used

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+var parts = [];
+var activityTypeCounts = _.countBy(activities, 'type');
+if (activityTypeCounts.asset_comment) {
+  if (activityTypeCounts.asset_comment === 1) {
```
This should be:

- 1 comment on 1 asset: `a new comment on an asset`
- multiple comments on 1 asset: `x comments on 1 asset`
- multiple comments on multiple assets: `x comments on y assets`

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+    parts.push('new comments on ' + activityTypeCounts.asset_comment + ' assets')
+  }
+}
+if (activityTypeCounts.asset_comment_reply) {
```
This should be:

- 1 reply on 1 comment: `a new reply on a comment`
- multiple replies on 1 comment: `x replies on 1 comment`
- multiple replies on multiple comments: `x replies on y comments`

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+  }
+}
+if (activityTypeCounts.whiteboard_chat) {
+  if (activityTypeCounts.whiteboard_chat === 1) {
```
This should be:

- 1 chat message on 1 whiteboard: `a new chat message on a whiteboard`
- multiple chat messages on 1 whiteboard: `x chat messages on 1 whiteboard`
- multiple chat messages on multiple whiteboards: `x chat messages on y whiteboards`

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+} else if (parts.length == 2) {
+  greeting += parts.join(' and ');
+} else {
+  greeting += parts[0] + ',' + parts[1] + ' and ' + parts[2]
```
semicolon

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+  }
+}
+
+if (parts.length == 1) {
```
triple equals

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+
+if (parts.length == 1) {
+  greeting += parts[0]
+} else if (parts.length == 2) {
```
triple equals

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+  greeting += parts[0]
+} else if (parts.length == 2) {
+  greeting += parts.join(' and ');
+} else {
```
Might either want to either

- make this an explicit `=== 3` or 
- join everything except the last item + `and` the last item

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+greeting += '.';
+
+// Only show a greeting if there is more than 1 activity type
+if (parts.length > 1) {
```
The greeting should be shown when there is more than 1 block in the email. For example, when an email has a block for a comment on an asset and second block for another comment on another asset, the greeting should be shown.

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                    var commenters = _.pluck(activity.asset.comments, 'user');
+                    var summaryActors = getSummaryActors(commenters, user);
+                -%>
+                    <%= summaryActors %> <a href="<%= assetLink %>">commented</a> on your asset ‘<a href="<%= assetLink %>"><%= activity.asset.title %></a>’.
```
commented doesn't have to be a link. The asset title is already a link

nicolaasmatthijs replied:
Should have double quotes around asset title link

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                -%>
+                    <%= summaryActors %> <a href="<%= assetLink %>">commented</a> on your asset ‘<a href="<%= assetLink %>"><%= activity.asset.title %></a>’.
+                <% } else if (activity.type === 'asset_comment_reply') {
+                    var assetLink = course.assetlibrary_url + '?col_asset=' + activity.asset.id
```
This block is the same as the one under `if (activity.type === 'asset_comment') {`

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                    var commenters = _.pluck(activity.asset.comments, 'user');
+                    var summaryActors = getSummaryActors(commenters, user);
+                -%>
+                    <%= summaryActors %> <a href="<%= assetLink %>">replied</a> to your comment on the asset ‘<a href="<%= assetLink %>"><%= activity.asset.title %></a>’.
```
replied doesn't have to be a link

nicolaasmatthijs replied:
Should have double quotes around asset title link

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                    var summaryActors = getSummaryActors(commenters, user);
+                -%>
+                    <%= summaryActors %> <a href="<%= assetLink %>">commented</a> on your asset ‘<a href="<%= assetLink %>"><%= activity.asset.title %></a>’.
+                <% } else if (activity.type === 'asset_comment_reply') {
```
When someone replies to your comment on your own asset, which summary is going to win?

simong replied:
I changed this so it will come out as a reply

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                    var whiteboardLink = course.whiteboards_url + '?col_whiteboard=' + activity.whiteboard.id
+                    if (activity.whiteboard.chats.length === 1) {
+                -%>
+                      There has been <a href="<%= whiteboardLink %>">1 new chat message</a> on your whiteboard ‘<a href="<%= whiteboardLink %>"><%= activity.whiteboard.title %></a>’.
```
`1 new chat message` doesn't have to be a link

nicolaasmatthijs replied:
Should have double quotes around whiteboard name

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                -%>
+                      There has been <a href="<%= whiteboardLink %>">1 new chat message</a> on your whiteboard ‘<a href="<%= whiteboardLink %>"><%= activity.whiteboard.title %></a>’.
+                    <% } else { -%>
+                      There have been <a href="<%= whiteboardLink %>"><%= activity.whiteboard.chats.length %> new chat messages</a> on your whiteboard ‘<a href="<%= whiteboardLink %>"><%= activity.whiteboard.title %></a>’.
```
`x new chat messages` doesn't have to be a link

nicolaasmatthijs replied:
Should have double quotes around whiteboard name

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                <tr>
+
+                  <%
+                      var thumbnailUrl = '';
```
You've switched to an indentation of 4 here

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                      var title = '';
+                      var thumbnailSummary = '';
+                      if (activity.asset) {
+                          link = course.assetlibrary_url + '?col_asset=' + activity.asset.id
```
semicolon

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                          link = course.assetlibrary_url + '?col_asset=' + activity.asset.id
+                          title = activity.asset.title;
+                          thumbnailSummary = 'by ';
+                          var owners = _.values(activity.asset.users)
```
semicolon

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                          var owners = _.values(activity.asset.users)
+                          var nrOfOwners = owners.length;
+                          if (nrOfOwners === 1) {
+                              thumbnailSummary += owners[0].canvas_full_name
```
semicolon

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                      } else if (activity.whiteboard) {
+                          link = course.whiteboards_url + '?col_whiteboard=' + activity.whiteboard.id;
+                          title = activity.whiteboard.title;
+                          thumbnailSummary = activity.whiteboard.users.length + ' collaborators';
```
Should be `1 collaborator` or `x collaborators`

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                          if (activity.whiteboard.thumbnail_url) {
+                              thumbnailUrl = activity.whiteboard.thumbnail_url;
+                          } else {
+                              thumbnailUrl = baseCollabosphereUrl + '/assets/img/icon_whiteboard.png'
```
semicolon

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                            <div class="preview-tile-metadata">
+                              <table cellpadding="0" cellspacing="0" class="preview-tile-metadata-inner">
+                                <tr valign="top">
+                                  <td colspan="2" class="preview-tile-metadata-title">
```
Why does this have a colspan of 2 ?

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                <div class="activity-comment">
+                  <a href="<%= authorLink %>" class="wrapped"><%= comment.user.canvas_full_name %></a>
+                  <span class="activity-date">on <%= formatDate(comment.created_at) %></span>
+                  <div class="activity-comment-message wrapped"><%- escape(comment.body).replace(/\n/g, '<br />') %></div>
```
URLs in comments should be converted to links

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+
+              oldestMessages.forEach(function(chat) {
+                var authorLink = course.assetlibrary_url + '?col_user=' + chat.user.id;
+                var authorThumbnail = chat.user.canvas_image;
```
Need to fall back to default profile picture here

nicolaasmatthijs commented on email-templates/daily/html.ejs:
```diff
@@ -0,0 +1,271 @@
+                <div class="activity-comment">
+                  <a href="<%= authorLink %>" class="wrapped"><%= chat.user.canvas_full_name %></a>
+                  <span class="activity-date">on <%= formatDate(chat.created_at) %></span>
+                  <div class="activity-comment-message wrapped"><%= chat.body %></div>
```
URLs in chat messages should be converted to links

nicolaasmatthijs commented on email-templates/whiteboard-invitation/html.ejs:
```diff
@@ -0,0 +1,99 @@
+
+            <div class="activity-summary-container">
+              <div class="wrapped">
+                You have been invited by <%= inviter.canvas_full_name %> to collaborate in the whiteboard ‘<a href="<%= whiteboardLink %>"><%= whiteboard.title %></a>’.
```
Should have double quotes around whiteboard title

nicolaasmatthijs commented on email-templates/whiteboard-invitation/html.ejs:
```diff
@@ -0,0 +1,99 @@
+                              <table cellpadding="0" cellspacing="0" class="preview-tile-metadata-inner">
+                                <tr valign="top">
+                                  <td colspan="2" class="preview-tile-metadata-title">
+                                    <a href="<%= whiteboardLink %>" title="<%= title %>" target="_blank" class="preview-tile-title wrapped"><%= title %></a>
```
No need to repeat title in title attribute

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+    // 2. Iterate through all the courses
+    async.eachSeries(courses, collectCourse, function(err) {
+      if (err) {
+          return log.error({'err': err}, 'Unable to retrieve the courses when sending notifications');
```
Error message should be `Unable to collect a course when sending notifications`

nicolaasmatthijs commented on node_modules/col-activities/lib/api.js:
```diff
@@ -35,6 +37,26 @@ var log = require('col-core/lib/logger')('col-activities');
+
+var diff = emailMoment.diff(now);
+log.info('Scheduling daily notifications for %s (in %d ms)', emailMoment.format(), diff);
+setTimeout(DailyNotifications.collect, diff);
```
I think the emails will only be sent once. I can't see anything that set the interval to send the emails out every day.

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+  getCourseData(course, function(err, data) {
+    if (err) {
+      // Don't pass the error up
+      emitter.emit(EVENT_NAMES.COURSE_COLLECTED, course);
```
Should we log the error at least?

simong replied:
All errors are logged in the implementation

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+/**
+ * Get the course data that had any activity in the last day. This will include:
+ *  - assets that were commented on
+ *  - assets were a comment was placed on a reply
```
typo

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+ * @param  {Object}     callback.err                  An error object, if any
+ * @param  {Object}     callback.data                 The relevant daily data in the course
+ * @param  {Object}     callback.data.users           All the users in the course (including those that weren't active)
+ * @param  {Object}     callback.data.assets          Assets with recent comments
```
Might want to specify that the recent comments are also included in this

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+ * @param  {Object}     callback.data                 The relevant daily data in the course
+ * @param  {Object}     callback.data.users           All the users in the course (including those that weren't active)
+ * @param  {Object}     callback.data.assets          Assets with recent comments
+ * @param  {Object}     callback.data.whiteboards     Whiteboards with recent chat messages
```
Might want to specify that the recent chat messages are also included in this

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+ * Get the assets which were commented on recently
+ *
+ * @param  {Context}    ctx                 Standard context containing the current user and the current course
+ * @param  {Object}     users               The users in the course (keyed by their user id)
```
It doesn't look like the assets are being used here

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+  };
+  DB.Asset.findAll(options).complete(function(err, assets) {
+    if (err) {
+      log.error({'err': err, 'course': ctx.course.id}, 'Unable to get the assets for a course');
```
`... get the assets with recent comments for a course`

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+      // Index the asset owners by their ID for quick lookups
+      asset.users = _.indexBy(asset.users, 'id');
+
+      // Add the commenters their user object
```
`Add the user objects for the commenters`

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+  };
+  DB.Whiteboard.findAll(queryOptions).complete(function(err, whiteboards) {
+    if (err) {
+      log.error({'err': err, 'course': ctx.course.id}, 'Unable to get the whiteboards for a course');
```
`the whiteboards with recent chat messages for a course`

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+        'model': DB.Chat,
+        'where': {
+          'created_at': {
+            'gt': yesterday
```
There's the potential for a chat message to be included in multiple emails as an email will include everything between 8 the previous day and the moment at which this is run (which will be a bit after 8)

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+/**
+ * Get a formatted timestamp for 24 hours ago
+ *
+ * @return {String}   A formatted timestamp that is suitable to use when querying through Sequelize
```
`@api private`

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+ * @return {String}   A formatted timestamp that is suitable to use when querying through Sequelize
+ */
+var get24HoursAgoTimestamp = function() {
+  return moment(new Date(Date.now() - 24 * 60 * 60 * 1000)).format('YYYY-MM-DD HH:mm:ss');
```
I think you can just do `moment().subtract(1, 'days').format(...`

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+      return callback();
+    }
+
+    // If there's no course data than we can return early
```
typo

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+      return callback();
+    }
+
+    // 2. Iterate through all the users in this course and send each user an e-mail
```
Comment is inaccurate. We only send emails to users that have updates

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+var handleUser = function(course, data, user, callback) {
+  // Ignore users that don't have an email address
+  if (!user.canvas_email) {
+      return callback();
```
Indentation

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+    });
+
+    // All replies made to the current user
+    var replies = _.filter(asset.comments, function(comment) {
```
I don't fully understand why we need to distinguish between `replies` and `otherReplies `. The only difference is that `replies` will contain the user's replies to their own comments, which I think is something that can always be excluded from the emails.

Therefore, I think we can just do 

```
var replies = _.filter(asset.comments, function(comment) {
  return (comment.parent && comment.parent.user_id === user.id && comment.user_id !== user.id);
});
```

and work with that

simong replied:
If we exclude your own replies, you might loose some context.

Assume the following comment chain:

> Mary: This picture demonstrates A
>   - Jack: No, it demonstrates B better
>   - Mary: But what if the sky is blue and water is wet? Then surely it must be A
>   - Jack: Ah yes, you're quite right, I had forgotten that water is wet

Mary's email would look like:
> Mary: This picture demonstrates A
>   - Jack: No, it demonstrates B better
>   - Jack: Ah yes, you're quite right, I had forgotten that water is wet

The distinction between `replies` and `otherReplies` was done so we're sure another user replied to your comment

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,562 @@
+    }
+  });
+
+  // Add whiteboards that the user is a member of which has new chat messages
```
`Add whiteboards that have new chat messages and the user is a member of`

simong commented:
I started following-up before I realized you hadn't finished your review. I've addressed most of the review up to now except for the quotes and links.

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+/**
+ * Copyright ©2015. The Regents of the University of California (Regents). All Rights Reserved.
```
All of the copyright notices can be changed to 2016

nicolaasmatthijs replied:
But maybe after the migration script PR is merged, which takes care of a large number of them

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+  // 1. Get all the active courses
+  CourseAPI.getCourses(false, function(err, courses) {
+    if (err) {
+        return log.error({'err': err}, 'Unable to retrieve the courses when sending notifications');
```
There seems to be some inconsistent spacing here

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+          return log.error({'err': err}, 'Unable to collect a course when sending notifications');
+      }
+
+      log.info('Sent emails for all courses, took %d ms', Date.now() - start);
```
It looks like we mostly do the logging along the following lines:

```
log.info({'duration': (Date.now() - start)}, 'Sent emails for all courses');
```

So it might be worth doing that here as well

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+  getCourseData(course, function(err, data) {
+    if (err) {
+      // Don't pass the error up
+      return callback();
```
Why doesn't this pass the error back up? Don't we want the `'Unable to collect a course when sending notifications'` error message in the logs ?

simong replied:
No, because if one course fails, all further courses would be skipped.

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+    // 2. Iterate through all the users in this course and notify them if necessary
+    var userIterator = handleUser.bind(null, course, data);
+    async.eachSeries(data.users, userIterator, function() {
+      log.info('Sent emails for course %d, took %d ms', course.id, Date.now() - start);
```
Same here

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+ */
+var getSubject = function(course, activities, currentUser) {
+  var subject = null;
+  if (activities.length === 1) {
```
Maybe add a line of docs here ?

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+  if (activities.length === 1) {
+    var activity = activities[0];
+    if (activity.type === 'asset_comment') {
+      var commenters = _.pluck(activity.asset.comments, 'user');
```
Note that extracting the `commenters` and `summaryActors` is the same in this block and the block below

nicolaasmatthijs replied:
It's actually pretty much the same in the third block as well

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+      if (comment.length > 40) {
+        comment = comment.substring(0, 40) + '...';
+      }
+      subject = summaryActors + ' replied to your comment ‘' + comment + '’';
```
The comment itself doesn't have to be part of the subject. It can just be `summaryActors + ' replied to your comment';`

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+  // If the canvas instance for this course has no name, we change the subject slightly as the
+  // "from name" already contains "Asset Library and Whiteboards"
+  } else {
+    subject = 'Today\'s activity in the Asset Library and Whiteboards';
```
This can be changed to:

- Both Asset Library and Whiteboards activity: `New Asset Library and Whiteboard activity is waiting for you`
- Only Asset Library activity: `New Asset Library activity is waiting for you`
- Only Whiteboards activity: `New Whiteboard activity is waiting for you`
- No Canvas name:  `New activity is waiting for you`

simong replied:
That last one only happens if there are no activities. If that's the case and we get this far, there's something wrong ^^

nicolaasmatthijs replied:
/facepalm/

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+/**
+ * An Asset activity
+ *
+ * @param {Asset}       asset       The asset on which there's some activity
```
There should be 2 spaces after `@param`

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+  asset = _.extend({}, asset);
+
+  if (type === 'asset_comment_reply') {
+    asset.comments = replies;
```
By setting `asset.comments` to `replies`, we're back to modifying the original asset (which is what the _.extend is trying to avoid)

simong replied:
No, the passed in `asset` object should not be changed

```javascript
$ node
> var lodash = require('lodash');
undefined
> var asset = {'comments': [1, 2, 3, 4, 5]}
undefined
> var ptr = asset // Create a pointer to the original object
undefined
> asset
{ comments: [ 1, 2, 3, 4, 5 ] }
> ptr
{ comments: [ 1, 2, 3, 4, 5 ] }
> var replies = [11, 12]
undefined
> asset = lodash.extend({}, asset)    // extend will copy over the entries from asset to a new object. This will be returned and assigned to the asset variable.
{ comments: [ 1, 2, 3, 4, 5 ] }
> asset
{ comments: [ 1, 2, 3, 4, 5 ] }
> ptr // Note that ptr points to the ORIGINAL object, whilst asset is now a completely new object
{ comments: [ 1, 2, 3, 4, 5 ] }
> asset.comments = replies
[ 11, 12 ]
> asset
{ comments: [ 11, 12 ] }
> ptr
{ comments: [ 1, 2, 3, 4, 5 ] }
```

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+  var newComments = asset.comments.length;
+
+  // Flatten the comments
+  var comments = replies || asset.comments;
```
This is duplicative of `asset.comments = replies;`

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+  // Flatten the comments
+  var comments = replies || asset.comments;
+  var commentsById = {};
+  _.each(comments, function(comment) {
```
I'm a bit confused here. Why are we turning the comments into an object and then straight away converting it to an array of the values?

simong replied:
It's a way of copying the comments as the next step will start modifying them. If any comment is re-used across (users their) activities, it would mess up the level

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+
+  asset.comments = commentsTree;
+
+  var that = {
```
You can just do the return here

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+      .value();
+
+    // We need all the data since yesterday
+    var yesterday = moment().subtract(1, 'day').format('YYYY-MM-DD HH:mm:ss')
```
I think this has the potential of including certain activities in 2 emails and not including certain activities at all, especially once we have a large number of courses:

- An activity took place at 2 minutes past the hour, but this code is executed 5 minutes past the hour. The next day, this code is executed 1 minute after the hour. This means that it would be included into the first email and the second email
- This code is executed at 2 minutes after the hour and an activity takes place at 3 minutes after the hour. The activity is not included in the email. The next day, this code is executed at 4 minutes after the hour and the activity is again not included in the email.

We should probably base the yesterday and today timestamps on the actual configured email time.

nicolaasmatthijs commented on node_modules/col-activities/lib/notifications/daily.js:
```diff
@@ -0,0 +1,548 @@
+          chat.user = users[chat.user_id];
+          return chat;
+        })
+        .sortBy('id')
```
Can we not specify that the chat messages should be ordered by id in `queryOptions`

simong replied:
Force of habit, Sequelize didn't used to support ordering on a joined table

nicolaasmatthijs commented on node_modules/col-activities/tests/test-daily.js:
```diff
@@ -0,0 +1,788 @@
+  /**
+   * Test that verifies that the email subject depends on which activities took place
+   */
+  it('changes the email subject depending on which activities took place', function(callback) {
```
This will need to be updated based on the review above

nicolaasmatthijs commented on node_modules/col-activities/tests/test-daily.js:
```diff
@@ -0,0 +1,788 @@
+          // Simon should get an email notifying him of a reply
+          ActivitiesTestUtil.assertDailyNotifications(dbCourse, [users.simon.me], function(emails) {
+            var simonsEmail = getEmail(emails, users.simon.me.canvas_email);
+            assert.strictEqual(simonsEmail.email.subject, users.nico.me.canvas_full_name + ' replied to your comment ‘This is a comment’');
```
Is there a way to verify that replies on the same comment are grouped together under that comment ?

nicolaasmatthijs replied:
And whatever else is necessary to validate the comment order?

nicolaasmatthijs commented on node_modules/col-activities/tests/test-daily.js:
```diff
@@ -0,0 +1,788 @@
+   * @throws {AssertionError}                       Error thrown when an assertion failed
+   * @api private
+   */
+  var getCourse = function(canvasCourseId, callback) {
```
It might be worth just making this a function in the course API ?

nicolaasmatthijs commented on node_modules/col-activities/tests/test-daily.js:
```diff
@@ -0,0 +1,788 @@
+  /**
+   * Assert that an email contains references to (and only to) a set of users
+   *
+   * @param  {Object[]}   email     Collected email (as per ActivitiesTestUtil.assertDailyNotifications)
```
Spacing

nicolaasmatthijs commented on node_modules/col-activities/tests/test-daily.js:
```diff
@@ -0,0 +1,788 @@
+
+      // Assert a full link to the user's asset library is present
+      var expectedLink = getAssetLibraryBaseUrl(email.course) + '?col_user=' + user.id;
+      assert.ok(email.email.html.indexOf(expectedLink) > -1);
```
It might be worth creating a `ensurePresent(string, fragment)` function to avoid doing the `.indexOf() > -1` everywhere

nicolaasmatthijs commented on node_modules/col-activities/tests/test-daily.js:
```diff
@@ -0,0 +1,788 @@
+  /**
+   * Get the tool URL for the asset library
+   *
+   * @param  {Course} course The course for which to get the asset library URL
```
Needs more spacing around `course` here and in the function below

nicolaasmatthijs commented on node_modules/col-activities/tests/test-daily.js:
```diff
@@ -0,0 +1,788 @@
+   * @api private
+   */
+  var getAssetLibraryBaseUrl = function(course) {
+    return getCourseUrl(course) + '/assetlibrary';
```
Isn't a tool's URL always `/external_tools/<toolid>` instead of `/external_tools/assetlibrary` ?

simong replied:
It is, but it's easier to debug this way

nicolaasmatthijs commented on node_modules/col-activities/tests/test-daily.js:
```diff
@@ -0,0 +1,788 @@
+   * Create a comment and reply to it
+   *
+   * @param  {Course}       course            The course in which to comment and reply on an asset
+   * @param  {Asset}        asset             The on asset on which to comment and reply
```
typo

nicolaasmatthijs commented on node_modules/col-activities/tests/util.js:
```diff
@@ -731,3 +732,61 @@ var assertEditActivityTypeConfigurationFails = module.exports.assertEditActivity
+    EmailUtil.removeListener(EmailUtil.EVENT_NAMES.EMAIL_SENT, emailListener);
+
+    // Verify the notifications were sent to the correct users
+    if (users) {
```
If no users are provided, shouldn't this verify that no emails were sent ?

nicolaasmatthijs commented on node_modules/col-core/lib/db.js:
```diff
@@ -242,6 +263,10 @@ var setUpModel = function(sequelize) {
       'type': Sequelize.STRING,
       'allowNull': true
     },
+    'canvas_email': {
```
Should the poller be storing the email address or do we want to wait for an explicit LTI launch for every user we send an email to ?

simong replied:
I missed that you can include the email address when requesting data from the API. I've added it to the poller. That way people who haven't launched into the LTI tool who are added to a whiteboard, will get an email when a chat message is added to it

nicolaasmatthijs commented on node_modules/col-core/lib/email.js:
```diff
@@ -0,0 +1,183 @@
+var config = require('config');
+var ejs = require('ejs');
+var EventEmitter = require('events');
+var EmailTemplate = require('email-templates').EmailTemplate
```
semicolon

nicolaasmatthijs commented on node_modules/col-core/lib/email.js:
```diff
@@ -0,0 +1,183 @@
+var cachedEmailTemplates = {};
+
+var emailEnabled = config.get('email.enabled');
+var sendgridClient = sendgrid(config.get('email.apiKey'));
```
Is this going to throw an error when no email api key is configured ?

nicolaasmatthijs commented on node_modules/col-core/lib/email.js:
```diff
@@ -0,0 +1,183 @@
+ *
+ * The email will be sent with the following from headers:
+ *  - from email address: noreply@<the course' canvas instance api domain>
+ *  - from name: The course canvas instance name or "The Asset Library and Whiteboards" if there's no configured name
```
It might be better to make this the name of the course instead of the name of the Canvas instance? There's more courses using it now.

nicolaasmatthijs commented on node_modules/col-core/lib/email.js:
```diff
@@ -0,0 +1,183 @@
+
+  // Otherwise we send the email through SendGrid's API
+  } else {
+    var email = new sendgridClient.Email(emailData);
```
Should we log the sending of the email here as well ?

nicolaasmatthijs commented on node_modules/col-core/lib/email.js:
```diff
@@ -0,0 +1,183 @@
+
+  // If we're not configured to send emails we log them
+  if (!emailEnabled) {
+    log.info(emailData, 'Sent an email');
```
Might want to make it clearer that it didn't actually send an email here

nicolaasmatthijs commented on node_modules/col-course/lib/api.js:
```diff
@@ -31,22 +31,30 @@ var log = require('col-core/lib/logger')('col-course');
+ * @param  {Number}       canvasCourseId                    The id of the course in Canvas
+ * @param  {Canvas}       canvas                            The Canvas instance the course is running on
+ * @param  {Object}       courseInfo                        Additional info for the course
+ * @param  {String}       [courseInfo.assetlibrary_url]     The URL where the asset library in this course can be reached
```
We should also keep track of the course name for the email from

nicolaasmatthijs commented on node_modules/col-course/lib/api.js:
```diff
@@ -31,22 +31,30 @@ var log = require('col-core/lib/logger')('col-course');
+    'courseInfo': Joi.object().keys({
+      'assetlibrary_url': Joi.string().optional(),
+      'whiteboards_url': Joi.string().optional()
+    }).required()
```
Is there a reason to make `courseInfo` required ?

simong replied:
Will there ever be a time where a course is created without an asset library / whiteboard url?

nicolaasmatthijs replied:
What if the engagement index is the first tool that's loaded in a course?

nicolaasmatthijs commented on node_modules/col-course/lib/api.js:
```diff
@@ -73,21 +83,37 @@ var getOrCreateCourse = module.exports.getOrCreateCourse = function(canvasCourse
  * @param  {Course[]}   callback.courses        The courses in the database. The associated `Canvas` object will be retrieved as well
  */
-var getCourses = module.exports.getCourses = function(callback) {
+var getCourses = module.exports.getCourses = function(includeInactive, callback) {
```
In what cases do the inactive courses need to be included ?

nicolaasmatthijs commented on package.json:
```diff
@@ -76,6 +79,7 @@
+    "sendgrid": "2.0.0",
     "sequelize": "3.7.1",
     "socket.io": "1.3.6",
     "validator": "1.1.3",
```
Can you include an image for `cole2.uconline.edu` as well?

nicolaasmatthijs commented on node_modules/col-rest/lib/rest/lti.js:
```diff
@@ -103,7 +112,23 @@ module.exports = function(client) {
+   */
+  var getRefererHeader = function(course, tool) {
+    var protocol = (course.canvas.use_https) ? 'https' : 'http';
+    return protocol + '://' + course.canvas.canvas_api_domain + '/external_tools/' + tool;
```
I'm still not entirely clear on how a `course.canvas.canvas_api_domain + '/external_tools/' + 'assetlibrary'` tool is being used as that results in an invalid Canvas URL

simong replied:
This header is only used during the tests. When you launch an LTI tool, LTI doesn't actually give you the parent URL of where that tool is being loaded from (which is why we look at the referer in the first place). This URL is missing the `/course/<id>` portion, which I'll add in the follow-up.

It doesn't really matter what this URL is for the tests, as long as it's different from the whiteboard tool URL. 

`protocol + '://' + course.canvas.canvas_api_domain + '/courses/' + course.id + '/external_tools/' + tool;`

nicolaasmatthijs commented on node_modules/col-users/lib/api.js:
```diff
@@ -65,7 +65,7 @@ var getUser = module.exports.getUser = function(id, callback) {
         'model': DB.Canvas,
         'as': 'canvas',
-        'attributes': ['canvas_api_domain', 'use_https']
+        'attributes': ['canvas_api_domain', 'use_https', 'logo', 'name']
```
Might be worth creating a `BASIC_CANVAS` property list to use every time a basic canvas instance needs to be retrieved

nicolaasmatthijs commented on node_modules/col-users/lib/constants.js:
```diff
@@ -24,6 +24,6 @@
 
 var Constants = module.exports = {
-  'BASIC_USER_FIELDS': ['id', 'canvas_user_id', 'canvas_course_role', 'canvas_enrollment_state', 'canvas_full_name', 'canvas_image'],
+  'BASIC_USER_FIELDS': ['id', 'canvas_user_id', 'canvas_course_role', 'canvas_enrollment_state', 'canvas_full_name', 'canvas_image', 'canvas_email'],
```
Does this mean that we're potentially exposing the email address in comments, etc. ? We should have a test that ensures we're not exposing the email address to other users.

nicolaasmatthijs commented on node_modules/col-whiteboards/lib/api.js:
```diff
@@ -400,26 +416,47 @@ var editWhiteboard = module.exports.editWhiteboard = function(ctx, id, title, me
+              callback(null, whiteboardProfile);
+
+              // Get the new members
+              oldMembers = _.indexBy(oldMembers, 'id');
```
You could do an _.pluck on the new and old members and then an _.without

simong replied:
We still need the full user objects, I think this is slightly easier to ready (and minimally better performant)

nicolaasmatthijs commented on node_modules/col-whiteboards/lib/api.js:
```diff
@@ -427,6 +464,42 @@ var editWhiteboard = module.exports.editWhiteboard = function(ctx, id, title, me
+      return callback();
+    }
+
+    var subject = util.format('You\'ve been invited to collaborate in the whiteboard ‘%s’', whiteboard.title);
```
`on the whiteboard`

nicolaasmatthijs commented on node_modules/col-whiteboards/tests/util.js:
```diff
@@ -92,6 +93,30 @@ var assertWhiteboard = module.exports.assertWhiteboard = function(whiteboard, op
+    // Because of the asynchronous process of sending out emails, we only return to the caller when
+    // both the REST request finishes and the emails have been sent out
+    var _whiteboard = null;
+    var done = _.after(2, function() {
```
Shouldn't this be the number of members (minus myself) rather than a hardcoded 2 ?

simong replied:
No, as we wait for:
 - The REST request to finish
 - The whiteboard api to emit INVITATIONS_SENT (which indicates all emails have been sent)

nicolaasmatthijs commented on node_modules/col-whiteboards/tests/util.js:
```diff
@@ -99,16 +124,37 @@ var assertCreateWhiteboard = module.exports.assertCreateWhiteboard = function(cl
+    // Require the whiteboards API inline as it requires the entire application to be bootstrapped
+    // when loaded the first time
+    var WhiteboardsAPI = require('../lib/api');
+    WhiteboardsAPI.once(WhiteboardsAPI.EVENT_NAMES.INVITATIONS_SENT, function() {
```
Wouldn't this miss a bug when multiple emails are accidentally sent out ?

simong replied:
It's hard to test that in the utility in a straightforward way, without leaking event listeners.

I'll add a test that:
 - attaches an event listener through `on` on the`EmailUtil` that expects an email for a user *once* and throws an error if there's more
 - create a whiteboard and add 1 user
 - edit a whiteboard and add 1 user

nicolaasmatthijs commented on node_modules/col-whiteboards/tests/util.js:
```diff
@@ -306,26 +352,82 @@ var assertGetWhiteboardsFails = module.exports.assertGetWhiteboardsFails = funct
  * @param  {String}             title                           The updated title of the whiteboard
  * @param  {Number[]}           members                         The ids of the users that should be a member of the whiteboard
  * @param  {Function}           callback                        Standard callback function
- * @param  {Whieboard}          callback.whiteboard             The updated whiteboard
```
A whiteboard sounds much better

![WiiiiBoard](http://cdn1-www.comingsoon.net/assets/uploads/2015/06/Back-to-the-Future-II-Hoverboard.jpg)

nicolaasmatthijs commented on node_modules/col-whiteboards/tests/util.js:
```diff
@@ -306,26 +352,82 @@ var assertGetWhiteboardsFails = module.exports.assertGetWhiteboardsFails = funct
+      // Because of the asynchronous process of sending out emails, we only return to the caller when
+      // both the REST request finishes and the emails have been sent out
+      var _whiteboard = null;
+      var done = _.after(2, function() {
```
Same here. Shouldn't this expect a number of emails equal to the number of new members ?

nicolaasmatthijs commented on node_modules/col-whiteboards/tests/util.js:
```diff
@@ -306,26 +352,82 @@ var assertGetWhiteboardsFails = module.exports.assertGetWhiteboardsFails = funct
+      // Require the whiteboards API inline as it requires the entire application to be bootstrapped
+      // when loaded the first time
+      var WhiteboardsAPI = require('../lib/api');
+      WhiteboardsAPI.once(WhiteboardsAPI.EVENT_NAMES.INVITATIONS_SENT, function() {
```
This is pretty much a complete repeat of what's above. Can it be consolidated?

nicolaasmatthijs commented:
[ ] Can you update the npm-shrinkwrap file to include the added dependencies?
[ ] Even though `https` is configured to `false`, it still looks like it's trying to load resources over `https` 

![screen shot 2016-01-21 at 6 49 15 pm](https://cloud.githubusercontent.com/assets/109850/12501213/b064921e-c06f-11e5-83f7-d814c2a7a671.png)

simong commented:
Pushed some follow-up

nicolaasmatthijs commented:
Left some final comments. Back to @simong

simong commented:
Pushed follow-up

nicolaasmatthijs commented:
Code looks good. Tests are failing though:

```
143 passing (45s)
  9 failing

  1) Daily activity emails sends emails to only those users who need to be notified:

      Uncaught AssertionError: 0 === 1
      + expected - actual

      -0
      +1
      
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/tests/util.js:767:12
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:89:14
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:461:16
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:588:12
      at null.<anonymous> (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-core/lib/globals/sequelize.js:69:16)
      at bound (domain.js:254:14)
      at runBound (domain.js:267:12)
      at tryCatcher (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:503:31)
      at Promise._settlePromiseAt (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:577:18)
      at Promise._settlePromises (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:693:14)
      at Async._drainQueue (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:123:16)
      at Async._drainQueues (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:133:10)
      at Immediate.Async.drainQueues [as _onImmediate] (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:15:14)
      at processImmediate [as _immediateCallback] (timers.js:358:17)

  2) Daily activity emails changes the email subject depending on which activities took place:

      Uncaught AssertionError: 0 === 6
      + expected - actual

      -0
      +6
      
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/tests/util.js:767:12
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:89:14
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:461:16
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:588:12
      at null.<anonymous> (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-core/lib/globals/sequelize.js:69:16)
      at bound (domain.js:254:14)
      at runBound (domain.js:267:12)
      at tryCatcher (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:503:31)
      at Promise._settlePromiseAt (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:577:18)
      at Promise._settlePromises (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:693:14)
      at Async._drainQueue (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:123:16)
      at Async._drainQueues (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:133:10)
      at Immediate.Async.drainQueues [as _onImmediate] (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:15:14)
      at processImmediate [as _immediateCallback] (timers.js:358:17)

  3) Daily activity emails groups replies on an asset owner's comment correctly:

      Uncaught AssertionError: 0 === 1
      + expected - actual

      -0
      +1
      
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/tests/util.js:767:12
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:89:14
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:461:16
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:588:12
      at null.<anonymous> (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-core/lib/globals/sequelize.js:69:16)
      at bound (domain.js:254:14)
      at runBound (domain.js:267:12)
      at tryCatcher (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:503:31)
      at Promise._settlePromiseAt (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:577:18)
      at Promise._settlePromises (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:693:14)
      at Async._drainQueue (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:123:16)
      at Async._drainQueues (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:133:10)
      at Immediate.Async.drainQueues [as _onImmediate] (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:15:14)
      at processImmediate [as _immediateCallback] (timers.js:358:17)

  4) Daily activity emails creates the correct greetings:

      Uncaught AssertionError: 0 === 1
      + expected - actual

      -0
      +1
      
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/tests/util.js:767:12
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:89:14
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:461:16
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:588:12
      at null.<anonymous> (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-core/lib/globals/sequelize.js:69:16)
      at bound (domain.js:254:14)
      at runBound (domain.js:267:12)
      at tryCatcher (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:503:31)
      at Promise._settlePromiseAt (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:577:18)
      at Promise._settlePromises (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:693:14)
      at Async._drainQueue (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:123:16)
      at Async._drainQueues (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:133:10)
      at Immediate.Async.drainQueues [as _onImmediate] (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:15:14)
      at processImmediate [as _immediateCallback] (timers.js:358:17)

  5) Daily activity emails should contain absolute links to the asset library and whiteboard tool:

      Uncaught AssertionError: 0 === 4
      + expected - actual

      -0
      +4
      
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/tests/util.js:767:12
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:89:14
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:461:16
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:588:12
      at null.<anonymous> (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-core/lib/globals/sequelize.js:69:16)
      at bound (domain.js:254:14)
      at runBound (domain.js:267:12)
      at tryCatcher (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:503:31)
      at Promise._settlePromiseAt (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:577:18)
      at Promise._settlePromises (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:693:14)
      at Async._drainQueue (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:123:16)
      at Async._drainQueues (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:133:10)
      at Immediate.Async.drainQueues [as _onImmediate] (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:15:14)
      at processImmediate [as _immediateCallback] (timers.js:358:17)

  6) Daily activity emails aggregates on asset or whiteboard:

      Uncaught AssertionError: 0 === 5
      + expected - actual

      -0
      +5
      
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/tests/util.js:767:12
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:89:14
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:461:16
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:588:12
      at null.<anonymous> (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-core/lib/globals/sequelize.js:69:16)
      at bound (domain.js:254:14)
      at runBound (domain.js:267:12)
      at tryCatcher (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:503:31)
      at Promise._settlePromiseAt (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:577:18)
      at Promise._settlePromises (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:693:14)
      at Async._drainQueue (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:123:16)
      at Async._drainQueues (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:133:10)
      at Immediate.Async.drainQueues [as _onImmediate] (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:15:14)
      at processImmediate [as _immediateCallback] (timers.js:358:17)

  7) Daily activity emails the amount of comments per activity is limited:

      Uncaught AssertionError: 0 === 1
      + expected - actual

      -0
      +1
      
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/tests/util.js:767:12
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:89:14
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:461:16
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:588:12
      at null.<anonymous> (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-core/lib/globals/sequelize.js:69:16)
      at bound (domain.js:254:14)
      at runBound (domain.js:267:12)
      at tryCatcher (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:503:31)
      at Promise._settlePromiseAt (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:577:18)
      at Promise._settlePromises (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:693:14)
      at Async._drainQueue (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:123:16)
      at Async._drainQueues (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:133:10)
      at Immediate.Async.drainQueues [as _onImmediate] (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:15:14)
      at processImmediate [as _immediateCallback] (timers.js:358:17)

  8) Daily activity emails the amount of chat messages per activity is limited:

      Uncaught AssertionError: 0 === 1
      + expected - actual

      -0
      +1
      
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/tests/util.js:767:12
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:89:14
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:461:16
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:588:12
      at null.<anonymous> (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-core/lib/globals/sequelize.js:69:16)
      at bound (domain.js:254:14)
      at runBound (domain.js:267:12)
      at tryCatcher (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:503:31)
      at Promise._settlePromiseAt (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:577:18)
      at Promise._settlePromises (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:693:14)
      at Async._drainQueue (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:123:16)
      at Async._drainQueues (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:133:10)
      at Immediate.Async.drainQueues [as _onImmediate] (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:15:14)
      at processImmediate [as _immediateCallback] (timers.js:358:17)

  9) Daily activity emails does not use old data:

      Uncaught AssertionError: 0 === 4
      + expected - actual

      -0
      +4
      
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/tests/util.js:767:12
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:89:14
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:461:16
      at /Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-activities/lib/notifications/daily.js:588:12
      at null.<anonymous> (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/col-core/lib/globals/sequelize.js:69:16)
      at bound (domain.js:254:14)
      at runBound (domain.js:267:12)
      at tryCatcher (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/util.js:26:23)
      at Promise._settlePromiseFromHandler (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:503:31)
      at Promise._settlePromiseAt (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:577:18)
      at Promise._settlePromises (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/promise.js:693:14)
      at Async._drainQueue (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:123:16)
      at Async._drainQueues (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:133:10)
      at Immediate.Async.drainQueues [as _onImmediate] (/Users/nicomatthijs/Desktop/ETS/Collabosphere/collabosphere/node_modules/sequelize/node_modules/bluebird/js/main/async.js:15:14)
      at processImmediate [as _immediateCallback] (timers.js:358:17)
```

simong commented:
This was due to sequalize badly handling string->date timezones

nicolaasmatthijs commented:
Merged!
